### PR TITLE
feat:  configurable post provisioning check + minor fixes

### PR DESCRIPTION
**Pull request by Tiboris**
feat: Make post provisioning ssh check configurable
    
Using post_provisioning_check with ssh default section
in the provisioning config to configure check based on
values in the dictionary:
```yaml
    post_provisioning_check:
        ssh:
            # Default configurations for every host
            enabled: True # True | False
            disabled_providers: ["podman"] # Per provider override to `enabled: True`
            enabled_providers: [] # Would be relevant if 'enabled' is 'False'
            # port: 22
            # timeout: 10 # minutes
    
            # Overrides

           group:
                ipaclient:
                    timeout: 309090 # minutes
            # If we want to override based on OS
            os:
                win-2012r2:
                    timeout: 15 # minutes
                win-2016:
                    timeout: 15 # minutes
                win-2019:
                    timeout: 15 # minutes
                fedora-34:
                    enabled: False
                    timeout: 1
                    enabled_providers: ["static"]
                    disabled_providers: ["beaker"]
```
Priority:
OS > Group > Default

### CODE REVIEW

**Comment by pvoborni** [comment]
Hi, thanks for the PR, I'm bit worried about limitations of the configuration regarding future enhancements and also that it is only about ssh. In general, SSH is not the only method how to run stuff on hosts (and thus pre-check that it is possible) and thus our checks should not limit to it. Example use case is mentioned in https://github.com/neoave/mrack/issues/142 . 

WDYT about the following config: 
```yaml

# defined in root of provisioning config, we don't need to mention defailt as by being in prof-config it becomes default behavior
connection_check:
  type: ssh # e.g. from options: ssh  | none, none would be the same as check: False
  # common options
  timeout: 10 # minutes
  # following type specific options,e.g.
  ssh_port: 22

# each provider can have it's own override, e.g.
# this also elimininate a need for exclusion list as 'none' will disable it for the provider
podman:
  connection_check: 
    type: none # or podman for native container connection 

```

WDYT?

**Comment by Tiboris** [comment]
Hello @pvoborni I have updated the code to match better suggestion, now it is ready to be reviewed.

**Comment by pvoborni** [comment]
We should probably create some kind of context object, e.g. in `up` action or transformer. Then pass all the relevant objects in it - the ones which are input and the ones which are ooutput of the operations. Then the code would be simpler and we would not need to enhance the Host object with additional data which are in metadata. 

E.g.

```python
class HostProvisioningContext:
    def __init__(self, name, req, meta_host):
        self.name = name
        self.req = req
        self.meta_host = meta_host
        self.spec = None
        self.result = None
        self.error = None

    @property
    def ok(self):
        """If provisioning resulted in success."""
        return self.error != None

```

**Comment by Tiboris** [comment]
Updated the code to move code to _check_ssh_auth()

**Review by pvoborni — COMMENTED** [review]
I was wondering why there is no reaction for the 2Q, and I had it in staging for couple of days.

File: src/mrack/providers/aws.py [hunk 1]
```diff
@@ -114,8 +114,13 @@ async def validate_hosts(self, reqs):
                         f"to use image: {req_img}"
                     )
 
+                try:  # FIXME when https://github.com/boto/boto3/issues/2531 fixed
+                    aws_img_name = aws_image.name
```
[the remark this answers is not in the document]
Comment (Tiboris) [reply src/mrack/providers/aws.py:118]: aws_image.name could be : ` Windows_Server-2016-English-Full-Base-2021.08.11` or similar the actual name of image.
The code is there for bettet readability, we can not tell what is the image based on ami

File: src/mrack/providers/provider.py [hunk 2]
```diff
@@ -142,6 +149,81 @@ async def _wait_for_ssh(self, host, port=22, timeout=1200):
 
         return res, host
 
+    async def _check_ssh_auth(self, default_check, active_hosts):
```
[the remark this answers is not in the document]
Comment (Tiboris) [reply src/mrack/providers/provider.py:152]: yes it's the only change @pvoborni

[the remark this answers is not in the document]
Comment (pvoborni) [reply src/mrack/providers/provider.py:152]: What about the plan for the unit test?

**Review by pvoborni — APPROVED** [review]
I'm approving as it is already tested in production.  

But I must note, that the PR is really not following good practices. And we should do better next time.
- combines fixes for independent problems and thus makes review more difficult - they could be merged independently faster
- no automated tests even though various pieces can be unit tested
- the lack of tests also doesn't motivate to structure code in a way that it can be unit tested

**Comment by Tiboris** [comment]
Thanks @pvoborni I will try to do better in future.

**Pull request merged**